### ISSUE TITLE
Consider supporting alternatives to threads for concurrency

### ISSUE THREAD

**Issue by gcbirzan** [GoogleCloudPlatform/google-cloud-python#4986]
Please, provide a way to use the pubsub library without having to spawn 5 threads.

**Comment by jonparrott** [comment]
@gcbirzan we did not make this decision lightly.

We picked threads because there are exactly two concurrency primitives available in Python  *2 and 3*: threads and processes. Processes are useful for dealing with CPU-bound tasks, whereas threads are useful for dealing with I/O bound tasks. All of our clients are I/O bound. With the additional context the grpcio uses threads by default and that threads are the most compatible and portable way to achieve concurrency (both gevent and asyncio have ways to co-operate with threads), it makes the most sense for us to use threads.

We are aware and sensitive to the fact that users have some aversion to threads, especially with environments like uwsgi which disable threads by default. We try to limit our usage of threads in clients that are meant to be used in a web-application- pub/sub's publisher, for example, only uses one background thread (other than the grpc threads) at a time. The subscriber, on the other hand, has a much different use case and requires multiple threads to accomplish its behavior.

With all that said, I am open to discussing issues with this approach and possibly alternatives. We *need* concurrency in some of the clients, so it's unlikely we can remove threads altogether, however, it's certainly possible for us to make changes to support other concurrency strategies at certain points. For example, we intentionally designed the pub/sub subscriber so that a different type of `Pool` could be used to execute tasks.
